crates/ruff/src/commands/format.rs

```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+            // annotate-snippets header. This is only an estimate because we don't actually know
+            // if the maximum line number present in the document will be rendered as part of
+            // the diff, either as a changed line or as an unchanged context line. For
+            // notebooks, we refine our estimate by checking the number of lines in each cell
+            // individually, otherwise we could use `formatted.source_code().count_lines(...)`
+            // in both cases.
+            let (fix, line_count) = if let SourceKind::IpyNotebook(formatted) = formatted
+                && let SourceKind::IpyNotebook(unformatted) = unformatted
+            {
+                notebook_index.insert(path.to_string(), unformatted.index().clone());
+
+                let mut edits = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .zip(unformatted.cell_offsets().ranges())
+                    .map(|(formatted_range, unformatted_range)| {
+                        let formatted = &formatted.source_code()[formatted_range];
+                        Edit::range_replacement(formatted.to_string(), unformatted_range)
+                    });
+
+                let fix = Fix::safe_edits(
+                    edits
+                        .next()
+                        .expect("Formatted files must have at least one edit"),
+                    edits,
+                );
+                let source = formatted.source_code();
+                let line_count = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .map(|range| source.count_lines(range))
+                    .max()
+                    .unwrap_or_default();
+                (fix, line_count)
+            } else {
+                let fix = Fix::safe_edit(Edit::range_replacement(
+                    formatted.source_code().to_string(),
+                    TextRange::up_to(unformatted.source_code().text_len()),
+                ));
+                let line_count = formatted
+                    .source_code()
+                    .count_lines(TextRange::up_to(formatted.source_code().text_len()));
+                (fix, line_count)
+            };
+
+            // Locate the first and last characters that differ to use as the diagnostic
+            // range.
+            let range = {
+                let mut start = None;
+                let mut end = None;
+                for ((offset, old), new) in unformatted
+                    .source_code()
+                    .char_indices()
+                    .zip(formatted.source_code().chars())
+                {
+                    if old != new {
+                        if start.is_none() {
+                            start = Some(offset);
+                        } else {
+                            end = Some(offset);
+                        }
+                    }
+                }
```

Comment:
I think we should use this range for the Edit as well if we already computed it anyway. It can help to reduce the size of the edit in common cases where only a few files are incorrectly formatted